solarforecastarbiter/datamodel.py

```diff
@@ -934,6 +938,227 @@ def __post_init__(self):
+    fill : str
+        Fill method to apply for times between those specified in `times`.
+        Options are 'forward' or 'backward'.
+    timezone : str, default None
+        IANA timezone string to use when constructing datetimes. If None,
+        the timezone of the observations is used, which is the report
+        timezone when calculated in a report.
+    """
+    times: Tuple[datetime.time, ...]
+    costs: Tuple[float, ...]
+    aggregation: str
+    net: bool
+    fill: str
+    timezone: str = None
+    __post_init__ = __validate_cost__('times')
+
+
+@dataclass(frozen=True)
+class DatetimeCost(BaseModel):
+    """Cost values based on datetimes.
+
+    Parameters
+    ----------
+    datetimes : tuple of datetime.datetime
+       The datetimes to associate with each cost value
+    costs : tuple of float
+       The cost per unit error of the forecasted variable for each datetime.
+       Must have the same length as `datetimes`.
+    aggregation : str
+        Aggregation method to use after calculating cost for the error series.
+        Currently only 'sum' or 'mean' are available.
+    net : bool
+        If True, compute the 'net' aggregate error instead of first calcuating
+        the absolute error before performing the aggregation.
+    fill : str
+       Fill method to apply for datetimes between those specified in
+       `datetimes`. Options are 'forward' or 'backward'.
+    timezone : str, default None
+        IANA timezone string to use when constructing datetimes. If None,
+        the timezone of the observations is used, which is the report
+        timezone when calculated in a report.    """
+    datetimes: Tuple[pd.Timestamp, ...]
+    costs: Tuple[float, ...]
+    aggregation: str
+    net: bool
+    fill: str
+    timezone: str = None
+    __post_init__ = __validate_cost__('datetimes')
+
+
+@dataclass(frozen=True)
+class ConstantCost(BaseModel):
+    """A constant cost per unit error of the forecasted variable
+
+    Parameters
+    ----------
+    cost : float
+    aggregation : str
+        Aggregation method to use after calculating cost for the error series.
+        Currently only 'sum' or 'mean' are available.
+    net : bool
+        If True, compute the 'net' aggregate error instead of first calcuating
+        the absolute error before performing the aggregation.
+    """
+    cost: float
+    aggregation: str
+    net: bool
+    __post_init__ = __validate_cost__(None)
+
+
+@dataclass(frozen=True)
+class CostBand(BaseModel):
+    """Cost specification for one error band
+
+    Parameters
+    ----------
+    error_range : tuple(float, float)
+        Bounds of the error to apply the specified cost function to.
+        Inf and -Inf are valid range points, and the error may be positive or
+        negative.
+    cost_function : str
+        One of 'timeofday', 'datetime', or 'constant'. Specifies which
+        cost model should be used to calculate the cost in this band.
```

Comment:
I haven't gotten too far yet, but this parameter seems unnecessary. Doesn't the type of `cost_function_parameters` tell you the same thing?

Reply:
or is this also needed for from_dict here?

Reply:
yes, but like for Cost, it's easier to read the dict/json and understand what type of cost the parameters are specifying. I don't think going off the keys/parameters (like we do for filters) is the right approach